src/rpc/src/chainStorageWatcher.ts

```diff
@@ -0,0 +1,167 @@
+import { AgoricChainStoragePathKind } from './types';
+import { batchVstorageQuery } from './batchQuery';
+import type { UpdateHandler } from './types';
+import type { Unserialize } from '@endo/marshal';
+
+type Subscriber<T> = {
+  onUpdate: UpdateHandler<T>;
+};
+
+const defaultNewPathQueryDelayMs = 20;
+const defaultRefreshLowerBoundMs = 2000;
+const defaultRefreshUpperBoundMs = 4000;
+
+const randomRefreshPeriod = (
+  refreshLowerBoundMs: number,
+  refreshUpperBoundMs: number,
+) =>
+  Math.floor(Math.random() * (refreshUpperBoundMs - refreshLowerBoundMs)) +
+  refreshLowerBoundMs;
+
+const makePathSubscriber = <T>(onUpdate: UpdateHandler<T>) => ({
+  onUpdate,
+});
+
+/**
+ * Periodically queries the most recent data from chain storage, batching RPC
+ * requests for efficiency.
+ *
+ * @param rpcAddr RPC server URL
+ * @param unserialize CapData unserializer to use
+ * @returns
+ */
+export const makeAgoricChainStorageWatcher = (
+  rpcAddr: string,
+  unserialize: Unserialize<string>,
+  newPathQueryDelayMs = defaultNewPathQueryDelayMs,
+  refreshLowerBoundMs = defaultRefreshLowerBoundMs,
+  refreshUpperBoundMs = defaultRefreshUpperBoundMs,
+) => {
+  // Map of paths to [identifier, value] pairs of most recent response values.
+  //
+  // The 'identifier' is used to avoid notifying subscribers for already-seen
+  // values. For 'data' queries, 'identifier' is the blockheight of the
+  // response. For 'children' queries, 'identifier' is the stringified array
+  // of children.
+  const latestValueCache = new Map<string, [string, unknown]>();
```

Comment:
the elements of the tuple can be named
```suggestion
  const latestValueCache = new Map<
    string,
    [identifier: string, value: unknown]
  >();
```